src/qutip_tensornetwork/core/data/network.py

```diff
@@ -672,3 +748,103 @@ def _match_edges_by_split(out_edges, in_edges):
     new_out_edges.reverse()
     new_in_edges.reverse()
     return new_out_edges, new_in_edges
+
+
+def _match_dimensions(edges, target_dims):
+    """This function splits the edges such that the dimensions of edges
+    matches target_dims. Only split of edges are allowed in order to match
+    target_dims.
+
+    Note
+    ----
+    The modifications to edges are performed in-place, which is why the
+    function is private as it is meant to be used only within network.
+
+    Examples
+    --------
+    >>> network_dim = [4]
+    >>> target_dims = [2, 2]
+    >>> array = np.random.random(network_dim)
+    >>> node = tn.Node(array)
+    >>> network = Network(node[:], [])
+    >>> _match_dimensions(network.in_edges, target_dims)
+
+    Since only splitting of edges are allowed, the next example will raise a
+    ValueError:
+    >>> network_dim = [2, 2]
+    >>> target_dims = [4]
+    >>> array = np.random.random(network_dim)
+    >>> node = tn.Node(array)
+    >>> network = Network(node[:], [])
+    >>> _match_dimensions(network.in_edges, target_dims) # ValueError
+    """
+
+    edges = edges[:]
+
+    target_dims = target_dims[:]
+    _target_dims = target_dims[:]  # This is kept for the error messages.
+
+    e_dims = [e.dimension for e in edges]
+
+    new_edges = []
```

Comment:
I would move the initialization of `new_edges` to just before the top of the loop so that it's closer to where it's used. Currently we define it, then don't use it for awhile, then suddenly use it in the middle of a loop.